File: leetcode/151_reverse_words_in_a_string/Solution.java
import java.util.*;

class Solution {
    public String reverseWords(String s) {
        String[] tokens = s.trim().split("\\s+");
        return String.join(" ", reverseStringArray(tokens));
    }

    List<String> reverseStringArray(String[] arr) {
        int len = arr.length;
        List<String> list = new ArrayList<>();
        for (int i = len-1; i >= 0; i--) {
            list.add(arr[i]);
        }
        return list;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.reverseWords("the sky is blue"));
        System.out.println(sol.reverseWords("a good   example"));
        System.out.println(sol.reverseWords("  hello world  "));
        System.out.println(sol.reverseWords("  Bob    Loves  Alice   "));
    }
}

File: leetcode/151_reverse_words_in_a_string/solution.py
class Solution:
    def reverseWords(self, s: str) -> str:
        import re
        split_tokens = re.split('\s+', s.strip())
        return ' '.join(split_tokens[::-1])


if __name__ == '__main__':
    sltn = Solution()
    assert sltn.reverseWords("the sky is blue") == "blue is sky the"
    assert sltn.reverseWords("a good   example") == "example good a"
    assert sltn.reverseWords("  hello world  ") == "world hello"
    assert sltn.reverseWords("  Bob    Loves  Alice   ") == "Alice Loves Bob"
